sdk/storage/src/clients/storage_account_client.rs

```diff
@@ -0,0 +1,427 @@
+    ) -> Arc<Box<Self>>
+    where
+        A: Into<String>,
+        S: AsRef<str>,
+    {
+        let account = account.into();
+
+        Arc::new(Box::new(Self {
+            blob_storage_uri: format!("https://{}.blob.core.windows.net", &account),
+            table_storage_uri: format!("https://{}.table.core.windows.net", &account),
+            queue_storage_uri: format!("https://{}.queue.core.windows.net", &account),
+            filesystem_uri: format!("https://{}.dfs.core.windows.net", &account),
+            storage_credentials: StorageCredentials::SASToken(get_sas_token_parms(
+                sas_token.as_ref(),
+            )),
+            http_client,
+        }))
+    }
+
+    pub fn http_client(&self) -> &dyn HttpClient {
+        self.http_client.as_ref().as_ref()
+    }
+
+    pub fn blob_storage_uri(&self) -> &str {
+        &self.blob_storage_uri
+    }
+
+    pub fn table_storage_uri(&self) -> &str {
+        &self.table_storage_uri
+    }
+    pub fn queue_storage_uri(&self) -> &str {
+        &self.queue_storage_uri
+    }
+
+    pub fn filesystem_uri(&self) -> &str {
+        &self.filesystem_uri
+    }
+
+    pub(crate) fn prepare_request<'a>(
+        &self,
+        url: &str,
+        method: &Method,
+        http_header_adder: &dyn Fn(Builder) -> Builder,
+        service_type: ServiceType,
+        request_body: Option<&'a [u8]>,
+    ) -> Result<(Request<&'a [u8]>, url::Url), AzureError> {
+        let dt = chrono::Utc::now();
+        let time = format!("{}", dt.format("%a, %d %h %Y %T GMT"));
+
+        let mut url = url::Url::parse(url)?;
+
+        // if we have a SAS token (in form of query pairs), let's add it to the url here
+        if let StorageCredentials::SASToken(query_pairs) = &self.storage_credentials {
+            for (k, v) in query_pairs {
+                url.query_pairs_mut().append_pair(k, v);
+            }
+        }
+
+        let mut request = Request::builder();
+        request = request.method(method).uri(url.as_str());
+
+        // let's add content length to avoid "chunking" errors.
+        match request_body {
+            Some(ref b) => request = request.header(CONTENT_LENGTH, &b.len().to_string() as &str),
+            None => request = request.header(CONTENT_LENGTH, "0"),
+        };
+
+        // This will give the caller the ability to add custom headers.
+        // The closure is needed to because request.headers_mut().set_raw(...) requires
+        // a Cow with 'static lifetime...
+        request = http_header_adder(request);
+
+        request = request
+            .header(MS_DATE, time)
+            .header(HEADER_VERSION, AZURE_VERSION);
+
+        // We sign the request only if it is not already signed (with the signature of an
+        // SAS token for example)
+        let request = match &self.storage_credentials {
+            StorageCredentials::Key(account, key) => {
+                if url.query_pairs().find(|p| p.0 == "sig").is_none() {
```

Comment:
Can you deconstruct the argument to closure so that you don't need to use `p.0`?